File: frontend/src/api/esg.ts
/**
 * ESG API client.
 *
 * Wraps /esg endpoints: overview, company detail, metrics CRUD, issues CRUD.
 *
 * The 3 pillars (E/S/G) shape the entire UI — most components display
 * tri-pillar splits (3-card strips, 3-column grids).
 */
import { api, type ModerationQueuedTag } from "./client";

export type Pillar = "E" | "S" | "G";
export type Severity = "low" | "med" | "high" | "critical";
export type IssueStatus = "open" | "in_progress" | "mitigated" | "closed";

export const PILLAR_META: { key: Pillar; label: string; color: string; full: string }[] = [
  { key: "E", label: "Окружающая среда", color: "#1D9E75", full: "Environmental" },
  { key: "S", label: "Социальная сфера", color: "#7F77DD", full: "Social" },
  { key: "G", label: "Корпуправление",   color: "#378ADD", full: "Governance" },
];

export const SEVERITY_META: { key: Severity; label: string; color: string }[] = [
  { key: "low",      label: "Низкая",       color: "#7DC4A0" },
  { key: "med",      label: "Средняя",      color: "#EF9F27" },
  { key: "high",     label: "Высокая",      color: "#E24B4A" },
  { key: "critical", label: "Критическая", color: "#991B1B" },
];

export const ISSUE_STATUS_META: { key: IssueStatus; label: string; color: string }[] = [
  { key: "open",        label: "Открыто",     color: "#E24B4A" },
  { key: "in_progress", label: "В работе",    color: "#EF9F27" },
  { key: "mitigated",   label: "Смягчено",    color: "#7DC4A0" },
  { key: "closed",      label: "Закрыто",     color: "#94A3B8" },
];

// ---------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------

export interface PillarStat {
  pillar: Pillar;
  metric_count: number;
  company_count: number;
  avg_target_attainment: number | null;
  avg_benchmark_diff: number | null;
  on_target_count: number;
  behind_count: number;
}

export interface IssueSeverityStat {
  severity: Severity;
  label: string;
  color: string;
  count: number;
}

export interface AgencyRatingCell {
  agency: string;
  rating_id: string | null;   // для inline-редактирования в таблице ESG
  rating: string | null;
  score: string | null;
  outlook: string | null;
  rating_date_text: string | null;
  report_url: string | null;
  is_recent: boolean;
}

export interface ESGCompanyScore {
  company_id: string;
  company_code: string;
  company_name: string | null;
  company_abbr: string | null;
  sector_code: string | null;
  sector_color: string | null;
  e_score: number | null;
  s_score: number | null;
  g_score: number | null;
  overall_score: number | null;
  metric_count: number;
  issues_open: number;
  issues_critical: number;
  last_year_reported: number | null;
  rank: number;
  ratings_by_agency: AgencyRatingCell[];
  composite_esg_score: number | null;     // 0..10 from agency ratings
  has_any_rating: boolean;
  recent_updates_count: number;
}

export interface ESGOverviewKpis {
  total_companies: number;
  companies_with_data: number;
  metrics_total: number;
  issues_open: number;
  issues_critical: number;
  avg_overall_score: number | null;
  covered_count: number;
  coverage_pct: number;
  leader_company_id: string | null;
  leader_company_name: string | null;
  leader_composite: number | null;
  leader_rating_letter: string | null;
  leader_ratings_count: number;
  unrated_count: number;
  recent_updates_count: number;
}

export interface AgencyCoverageStat {
  agency: string;
  count: number;
  color: string;
}

export interface RecentRatingUpdate {
  company_id: string;
  company_code: string;
  company_name: string;
  sector_code: string | null;
  sector_color: string | null;
  agency: string;
  agency_color: string;
  rating: string | null;
  score: string | null;
  rating_date_text: string | null;
  report_url: string | null;
}

export interface SectorBreakdownItem {
  code: string;
  label: string;
  color: string;
  total: number;
  covered: number;
  coverage_pct: number;
  leader_company_id: string | null;
  leader_company_name: string | null;
  leader_composite: number | null;
}

export interface ESGOverviewResponse {
  year: number | null;
  sector_code: string | null;
  kpis: ESGOverviewKpis;
  pillars: PillarStat[];
  issue_severity_split: IssueSeverityStat[];
  rankings: ESGCompanyScore[];
  agency_coverage: AgencyCoverageStat[];
  sector_breakdown: SectorBreakdownItem[];
  recent_updates: RecentRatingUpdate[];
  available_years: number[];
  sectors: { code: string; count: number }[];
  generated_at: string;
}

export interface ESGMetricBrief {
  id: string;
  company_id: string;
  year: number;
  pillar: Pillar;
  metric_code: string;
  metric_name: string;
  value: number | null;
  unit: string | null;
  target: number | null;
  benchmark: number | null;
  notes: string | null;
  target_attainment_pct: number | null;
  benchmark_diff_pct: number | null;
}

export interface ESGIssueBrief {
  id: string;
  company_id: string;
  company_code: string | null;
  company_name: string | null;
  pillar: Pillar;
  title: string;
  description: string | null;
  severity: Severity | null;
  status: IssueStatus;
  created_at: string;
}

export interface ESGCompanyDetail {
  company_id: string;
  company_code: string;
  company_name: string | null;
  sector_code: string | null;
  year: number;
  e_score: number | null;
  s_score: number | null;
  g_score: number | null;
  overall_score: number | null;
  metrics_e: ESGMetricBrief[];
  metrics_s: ESGMetricBrief[];
  metrics_g: ESGMetricBrief[];
  issues: ESGIssueBrief[];
  available_years: number[];
  tracked_years: number[];
}

export interface ESGMetricUpsertPayload {
  company_id: string;
  year: number;
  pillar: Pillar;
  metric_code: string;
  metric_name: string;
  value?: number | null;
  unit?: string | null;
  target?: number | null;
  benchmark?: number | null;
  notes?: string | null;
}

export interface ESGIssueCreatePayload {
  company_id: string;
  pillar: Pillar;
  title: string;
  description?: string | null;
  severity?: Severity;
}

export interface ESGIssueUpdatePayload {
  pillar?: Pillar;
  title?: string;
  description?: string | null;
  severity?: Severity;
  status?: IssueStatus;
}

// ---------------------------------------------------------------------
// ESG Maturity Cockpit — матрица зрелости + EMS
// ---------------------------------------------------------------------

export interface ESGMaturityCellBrief {
  dimension: string;
  sub_key: string;
  stage: number;
  status_text?: string | null;
  value_text?: string | null;
  evidence_url?: string | null;
  due_date?: string | null;
}
export interface ESGRatingMini {
  agency: string;
  rating: string | null;
  score: string | null;
  outlook: string | null;
  report_url: string | null;
}
export interface ESGMaturityCompany {
  company_id: string;
  company_code: string;
  company_name?: string | null;
  sector_code?: string | null;
  sector_name?: string | null;
  sector_color?: string | null;
  cells: ESGMaturityCellBrief[];
  dim_stage: Record<string, number>;   // D1..D6 → 0..4
  ems: number;                         // 0..100
  rating_count: number;
  ratings: ESGRatingMini[];            // сами ESG-рейтинги (агентство/значение/ссылка)
  not_needed?: boolean;                // «не нуждается» → исключена из метрик/статистики
}
export interface ESGMaturityHeatmap {
  year: number;
  companies: ESGMaturityCompany[];
  ems_mean: number;
  ems_median: number;
  ems_delta_yoy?: number | null;
  baskets: { mature: number; developing: number; starting: number };
  climate_funnel: number[];   // passed per stage 1..4
  risk_funnel: number[];      // passed per stage 1..3
  iso_full_count: number;
  rated_count: number;
  total_companies: number;
  available_years: number[];
  generated_at: string;
}
export interface ESGMaturityCellUpsertPayload {
  company_id: string;
  year: number;
  dimension: string;
  sub_key?: string;
  stage?: number | null;
  status_text?: string | null;
  value_text?: string | null;
  evidence_url?: string | null;
  due_date?: string | null;
  extra?: Record<string, unknown> | null;
}

// ---------------------------------------------------------------------
// ESG SWOT / выводы
// ---------------------------------------------------------------------
export interface ESGSwotItemBrief {
  id?: string | null;
  kind: "strength" | "weakness";
  scope: "portfolio" | "company";
  company_id?: string | null;
  company_code?: string | null;
  company_name?: string | null;
  title?: string | null;
  body: string;
  severity?: string | null;
  order_idx: number;
}
export interface ESGSwotResponse {
  portfolio_strengths: ESGSwotItemBrief[];
  portfolio_weaknesses: ESGSwotItemBrief[];
  company_items: ESGSwotItemBrief[];
  generated_at: string;
}
export interface ESGSwotUpsertPayload {
  id?: string | null;
  kind: "strength" | "weakness";
  scope: "portfolio" | "company";
  company_id?: string | null;
  title?: string | null;
  body: string;
  severity?: string | null;
  order_idx?: number;
}

// ---------------------------------------------------------------------
// ESG-отчёты по годам (годовая таблица в профиле зрелости)
// ---------------------------------------------------------------------
export interface ESGReportBrief {
  id?: string | null;
  company_id: string;
  year: number;
  status: string | null;
  report_url: string | null;
  note: string | null;
  changed_by_name: string | null;
  updated_at: string | null;
}
export interface ESGReportListResponse {
  company_id: string;
  company_code: string | null;
  company_name: string | null;
  items: ESGReportBrief[];
  last_changed_by_name: string | null;
  last_changed_at: string | null;
  last_changed_year: number | null;
  generated_at: string;
}
export interface ESGReportUpsertPayload {
  company_id: string;
  year: number;
  status?: string | null;
  report_url?: string | null;
  note?: string | null;
}

// ---------------------------------------------------------------------
// ESG-релевантные KPI по компаниям (подтягиваются из модуля KPI по контексту)
// ---------------------------------------------------------------------
export interface ESGKpiBrief {
  name: string;
  unit: string | null;
  manager: string | null;
  plan: number | null;
  fact: number | null;
  pct: number | null;
  direction: string;
}
export interface ESGKpiCompany {
  company_id: string;
  company_code: string | null;
  kpis: ESGKpiBrief[];
}
export interface ESGKpiResponse {
  year: number;
  items: ESGKpiCompany[];
  generated_at: string;
}
export interface ESGKpiCreatePayload {
  company_id: string;
  year: number;
  name: string;
  unit?: string | null;
  direction?: "up" | "down";
  plan?: number | null;
  fact?: number | null;
}

// ---------------------------------------------------------------------
// API
// ---------------------------------------------------------------------

export const esgApi = {
  async getOverview(params: { year?: number; sector_code?: string; rankings_limit?: number } = {}) {
    const r = await api.get<ESGOverviewResponse>("/esg/overview", { params });
    return r.data;
  },

  async getCompanyDetail(companyId: string, year?: number) {
    const r = await api.get<ESGCompanyDetail>(
      `/esg/companies/${companyId}`,
      { params: year ? { year } : {} },
    );
    return r.data;
  },

  async getMaturityHeatmap(year?: number): Promise<ESGMaturityHeatmap> {
    const r = await api.get<ESGMaturityHeatmap>("/esg/maturity/heatmap", { params: year ? { year } : {} });
    return r.data;
  },

  async upsertMaturityCell(payload: ESGMaturityCellUpsertPayload): Promise<ESGMaturityCellBrief | ModerationQueuedTag> {
    const r = await api.put<ESGMaturityCellBrief | ModerationQueuedTag>("/esg/maturity/cell", payload);
    return r.data;
  },

  async getSwot(): Promise<ESGSwotResponse> {
    const r = await api.get<ESGSwotResponse>("/esg/swot");
    return r.data;
  },

  async upsertSwot(payload: ESGSwotUpsertPayload): Promise<ESGSwotItemBrief | ModerationQueuedTag> {
    const r = await api.put<ESGSwotItemBrief | ModerationQueuedTag>("/esg/swot", payload);
    return r.data;
  },

  async getCompanyReports(companyId: string): Promise<ESGReportListResponse> {
    const r = await api.get<ESGReportListResponse>(`/esg/companies/${companyId}/reports`);
    return r.data;
  },

  async upsertReport(payload: ESGReportUpsertPayload): Promise<ESGReportBrief | ModerationQueuedTag> {
    const r = await api.put<ESGReportBrief | ModerationQueuedTag>("/esg/report", payload);
    return r.data;
  },

  async getEsgKpis(year?: number): Promise<ESGKpiResponse> {
    const r = await api.get<ESGKpiResponse>("/esg/kpis", { params: year ? { year } : {} });
    return r.data;
  },

  async addEsgKpi(payload: ESGKpiCreatePayload): Promise<ESGKpiBrief> {
    const r = await api.post<ESGKpiBrief>("/esg/kpi", payload);
    return r.data;
  },

  async upsertMetric(payload: ESGMetricUpsertPayload): Promise<ESGMetricBrief | ModerationQueuedTag> {
    const r = await api.put<ESGMetricBrief | ModerationQueuedTag>("/esg/metric", payload);
    return r.data;
  },

  async deleteMetric(metricId: string) {
    await api.delete(`/esg/metric/${metricId}`);
  },

  async listIssues(params: {
    company_id?: string;
    pillar?: Pillar;
    severity?: Severity;
    status?: IssueStatus;
    limit?: number;
  } = {}) {
    const r = await api.get<ESGIssueBrief[]>("/esg/issues", { params });
    return r.data;
  },

  async createIssue(payload: ESGIssueCreatePayload): Promise<ESGIssueBrief | ModerationQueuedTag> {
    const r = await api.post<ESGIssueBrief | ModerationQueuedTag>("/esg/issue", payload);
    return r.data;
  },

  async updateIssue(issueId: string, payload: ESGIssueUpdatePayload): Promise<ESGIssueBrief | ModerationQueuedTag> {
    const r = await api.patch<ESGIssueBrief | ModerationQueuedTag>(`/esg/issue/${issueId}`, payload);
    return r.data;
  },

  async deleteIssue(issueId: string) {
    await api.delete(`/esg/issue/${issueId}`);
  },
};

// ---------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------

export function pillarMeta(p: Pillar | string): { label: string; color: string } {
  const m = PILLAR_META.find((x) => x.key === p);
  return m ? { label: m.label, color: m.color } : { label: String(p), color: "#94A3B8" };
}

export function severityMeta(s: Severity | string | null | undefined): { label: string; color: string } {
  if (!s) return { label: "—", color: "#94A3B8" };
  const m = SEVERITY_META.find((x) => x.key === s);
  return m ? { label: m.label, color: m.color } : { label: String(s), color: "#94A3B8" };
}

export function issueStatusMeta(s: IssueStatus | string | null | undefined): { label: string; color: string } {
  if (!s) return { label: "—", color: "#94A3B8" };
  const m = ISSUE_STATUS_META.find((x) => x.key === s);
  return m ? { label: m.label, color: m.color } : { label: String(s), color: "#94A3B8" };
}

/** Score 0-100 → color: green ≥75, mint ≥50, amber ≥25, red <25. */
export function scoreColor(s: number | null | undefined): string {
  if (s == null) return "#94A3B8";
  if (s >= 75) return "#1D9E75";
  if (s >= 50) return "#7DC4A0";
  if (s >= 25) return "#EF9F27";
  return "#E24B4A";
}

/** Format ESG metric value with unit. */
export function fmtMetricValue(v: number | null | undefined, unit: string | null | undefined): string {
  if (v == null) return "—";
  const n = Number(v);
  if (isNaN(n)) return "—";
  const numStr = n.toLocaleString("ru-RU", { maximumFractionDigits: 2 });
  return unit ? `${numStr} ${unit}` : numStr;
}

File: backend/app/schemas/esg.py
"""ESG schemas — overview dashboard, per-company detail, metrics editing.

ESG model:
  - 3 pillars: E (environmental), S (social), G (governance)
  - Per-company per-year per-metric values, with optional target/benchmark
  - Material issues with severity (low/med/high/critical) and status (open/...)
  - Free-text notes tied to company/year/metric
  - Per-company year tracking (which years are actively reported)
"""
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas._types import MoneyDecimal

# =====================================================================
# Pillar aggregates
# =====================================================================

class PillarStat(BaseModel):
    """Per-pillar aggregates across the portfolio."""
    pillar: str                                    # E | S | G
    metric_count: int = 0
    company_count: int = 0
    avg_target_attainment: Optional[float] = None  # avg(value / target) when target present, %
    avg_benchmark_diff: Optional[float] = None     # avg deviation from benchmark, %
    on_target_count: int = 0                       # metrics where value >= target
    behind_count: int = 0                          # metrics where value < target


class IssueSeverityStat(BaseModel):
    severity: str
    label: str
    color: str
    count: int = 0


# =====================================================================
# Company-level rolled-up score
# =====================================================================

class AgencyRatingCell(BaseModel):
    """A single agency rating slot for ESGCompanyScore.ratings_by_agency.

    `rating=None` means the company has no rating from that agency yet.
    """
    agency: str
    rating_id: Optional[UUID] = None     # id рейтинга — для inline-редактирования в таблице ESG
    rating: Optional[str] = None
    score: Optional[str] = None
    outlook: Optional[str] = None
    rating_date_text: Optional[str] = None
    report_url: Optional[str] = None
    is_recent: bool = False              # updated within current or previous year


class ESGCompanyScore(BaseModel):
    """Per-company ESG snapshot for the rankings table."""
    company_id: UUID
    company_code: str
    company_name: Optional[str] = None
    company_abbr: Optional[str] = None
    sector_code: Optional[str] = None
    sector_color: Optional[str] = None

    e_score: Optional[float] = None     # 0..100 normalized score, E pillar
    s_score: Optional[float] = None
    g_score: Optional[float] = None
    overall_score: Optional[float] = None

    metric_count: int = 0
    issues_open: int = 0
    issues_critical: int = 0

    last_year_reported: Optional[int] = None
    rank: int = 0

    # ── Legacy-style agency ratings (Sustainable Fitch / S&P ESG / CDP / …) ──
    # Per-agency cell. Missing agency means no rating yet.
    ratings_by_agency: list[AgencyRatingCell] = Field(default_factory=list)
    # Composite (0..10) computed from agency ratings — legacy `_esgComposite`.
    composite_esg_score: Optional[float] = None
    has_any_rating: bool = False
    recent_updates_count: int = 0       # # of agency ratings updated recently


class AgencyCoverageStat(BaseModel):
    agency: str
    count: int
    color: str


# =====================================================================
# Issues
# =====================================================================

class ESGIssueBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    company_code: Optional[str] = None
    company_name: Optional[str] = None
    pillar: str
    title: str
    description: Optional[str] = None
    severity: Optional[str] = None
    status: str
    created_at: datetime


class ESGIssueCreate(BaseModel):
    company_id: UUID
    pillar: str = Field(..., pattern="^(E|S|G)$")
    title: str = Field(..., min_length=1, max_length=512)
    description: Optional[str] = None
    severity: str = Field("med", pattern="^(low|med|high|critical)$")


class ESGIssueUpdate(BaseModel):
    pillar: Optional[str] = Field(None, pattern="^(E|S|G)$")
    title: Optional[str] = Field(None, min_length=1, max_length=512)
    description: Optional[str] = None
    severity: Optional[str] = Field(None, pattern="^(low|med|high|critical)$")
    status: Optional[str] = Field(None, pattern="^(open|in_progress|mitigated|closed)$")


# =====================================================================
# Metrics (the editable rows for a single company)
# =====================================================================

class ESGMetricBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    year: int
    pillar: str
    metric_code: str
    metric_name: str
    value: Optional[MoneyDecimal] = None
    unit: Optional[str] = None
    target: Optional[MoneyDecimal] = None
    benchmark: Optional[MoneyDecimal] = None
    notes: Optional[str] = None

    target_attainment_pct: Optional[float] = None
    benchmark_diff_pct: Optional[float] = None


class ESGMetricUpsert(BaseModel):
    """Used for both create + update (PUT semantics)."""
    company_id: UUID
    year: int = Field(..., ge=2000, le=2100)
    pillar: str = Field(..., pattern="^(E|S|G)$")
    metric_code: str = Field(..., min_length=1, max_length=64)
    metric_name: str = Field(..., min_length=1, max_length=255)
    value: Optional[MoneyDecimal] = None
    unit: Optional[str] = Field(None, max_length=32)
    target: Optional[MoneyDecimal] = None
    benchmark: Optional[MoneyDecimal] = None
    notes: Optional[str] = None


# =====================================================================
# Company detail (drill view)
# =====================================================================

class ESGCompanyDetail(BaseModel):
    company_id: UUID
    company_code: str
    company_name: Optional[str] = None
    sector_code: Optional[str] = None
    year: int

    e_score: Optional[float] = None
    s_score: Optional[float] = None
    g_score: Optional[float] = None
    overall_score: Optional[float] = None

    metrics_e: list[ESGMetricBrief] = Field(default_factory=list)
    metrics_s: list[ESGMetricBrief] = Field(default_factory=list)
    metrics_g: list[ESGMetricBrief] = Field(default_factory=list)
    issues: list[ESGIssueBrief] = Field(default_factory=list)

    available_years: list[int] = Field(default_factory=list)
    tracked_years: list[int] = Field(default_factory=list)


# =====================================================================
# Overview response (the dashboard root)
# =====================================================================

class ESGOverviewKpis(BaseModel):
    total_companies: int = 0
    companies_with_data: int = 0
    metrics_total: int = 0
    issues_open: int = 0
    issues_critical: int = 0
    avg_overall_score: Optional[float] = None    # 0..100

    # ── Legacy-style KPI strip (4 cards) ──
    covered_count: int = 0               # companies with ≥1 agency rating
    coverage_pct: int = 0                # covered/total %
    planned_count: int = 0               # companies с отметкой «запланировано» (rp) без рейтинга
    leader_company_id: Optional[UUID] = None
    leader_company_name: Optional[str] = None
    leader_composite: Optional[float] = None     # 0..10
    leader_rating_letter: Optional[str] = None   # "AA", "BBB-", etc.
    leader_ratings_count: int = 0        # how many agencies rate the leader
    unrated_count: int = 0
    recent_updates_count: int = 0        # # of (co, agency) updates in current+previous year


class RecentRatingUpdate(BaseModel):
    """Legacy `recentUpdates` row: company × agency rating updated recently."""
    company_id: UUID
    company_code: str
    company_name: str
    sector_code: Optional[str] = None
    sector_color: Optional[str] = None
    agency: str
    agency_color: str
    rating: Optional[str] = None
    score: Optional[str] = None
    rating_date_text: Optional[str] = None
    report_url: Optional[str] = None


class SectorBreakdownItem(BaseModel):
    code: str
    label: str
    color: str
    total: int
    covered: int
    coverage_pct: int
    leader_company_id: Optional[UUID] = None
    leader_company_name: Optional[str] = None
    leader_composite: Optional[float] = None


class ESGOverviewResponse(BaseModel):
    year: Optional[int] = None
    sector_code: Optional[str] = None

    kpis: ESGOverviewKpis
    pillars: list[PillarStat] = Field(default_factory=list)
    issue_severity_split: list[IssueSeverityStat] = Field(default_factory=list)
    rankings: list[ESGCompanyScore] = Field(default_factory=list)

    # Legacy-style aggregates:
    agency_coverage: list[AgencyCoverageStat] = Field(default_factory=list)
    sector_breakdown: list[SectorBreakdownItem] = Field(default_factory=list)
    recent_updates: list[RecentRatingUpdate] = Field(default_factory=list)

    available_years: list[int] = Field(default_factory=list)
    sectors: list[dict] = Field(default_factory=list)

    generated_at: datetime


# =====================================================================
# ESG Maturity Cockpit — матрица зрелости 22×6 + EMS
# =====================================================================

class ESGMaturityCellBrief(BaseModel):
    dimension: str
    sub_key: str = ""
    stage: int = 0
    status_text: Optional[str] = None
    value_text: Optional[str] = None
    evidence_url: Optional[str] = None
    due_date: Optional[str] = None


class ESGRatingMini(BaseModel):
    """Компактный ESG-рейтинг для отображения/inline-правки прямо в матрице зрелости."""
    id: Optional[UUID] = None
    agency: str
    rating: Optional[str] = None
    score: Optional[str] = None
    outlook: Optional[str] = None
    report_url: Optional[str] = None
    prev: Optional[str] = None     # предыдущее значение (для динамики «старый → новый»)


class ESGMaturityCompany(BaseModel):
    company_id: UUID
    company_code: str
    company_name: Optional[str] = None
    sector_code: Optional[str] = None
    sector_name: Optional[str] = None
    sector_color: Optional[str] = None
    cells: list[ESGMaturityCellBrief] = Field(default_factory=list)
    dim_stage: dict[str, int] = Field(default_factory=dict)   # D1..D6 → 0..4
    ems: float = 0.0                                          # 0..100
    rating_count: int = 0
    ratings: list[ESGRatingMini] = Field(default_factory=list)   # сами ESG-рейтинги (агентство/значение/ссылка)
    not_needed: bool = False                                  # «не нуждается» → исключена из метрик/статистики
    dim_not_required: list[str] = Field(default_factory=list)    # измерения «не требуется» (D1..D5)
    # Кол-во загруженных документов по этапам ESG: {"D4:2": 3, "D1:iso14001": 1, …}
    # entity_id = "<dim>:<stageIdx>" (climate D4:1..4 / risk D5:1..3 / ISO D1:iso*).
    stage_doc_counts: dict[str, int] = Field(default_factory=dict)


class ESGMaturityBaskets(BaseModel):
    mature: int = 0       # EMS >= 70
    developing: int = 0   # 40..69
    starting: int = 0     # < 40


class ESGMaturityHeatmap(BaseModel):
    year: int
    companies: list[ESGMaturityCompany] = Field(default_factory=list)
    ems_mean: float = 0.0
    ems_median: float = 0.0
    ems_delta_yoy: Optional[float] = None
    baskets: ESGMaturityBaskets = Field(default_factory=ESGMaturityBaskets)
    climate_funnel: list[int] = Field(default_factory=list)   # passed per stage 1..4
    risk_funnel: list[int] = Field(default_factory=list)      # passed per stage 1..3
    iso_full_count: int = 0                                   # компаний со всеми 3 ISO
    rated_count: int = 0
    total_companies: int = 0
    available_years: list[int] = Field(default_factory=list)
    generated_at: datetime


class ESGMaturityCellUpsert(BaseModel):
    company_id: UUID
    year: int = Field(..., ge=2000, le=2100)
    dimension: str = Field(..., max_length=8)
    sub_key: str = Field("", max_length=32)
    stage: Optional[int] = Field(None, ge=0, le=4)
    status_text: Optional[str] = Field(None, max_length=64)
    value_text: Optional[str] = Field(None, max_length=255)
    evidence_url: Optional[str] = None
    due_date: Optional[str] = None
    extra: Optional[dict] = None


# =====================================================================
# ESG SWOT / выводы (портфель + по-компанийно)
# =====================================================================

class ESGSwotItemBrief(BaseModel):
    id: Optional[UUID] = None
    kind: str                       # strength | weakness
    scope: str = "portfolio"        # portfolio | company
    company_id: Optional[UUID] = None
    company_code: Optional[str] = None
    company_name: Optional[str] = None
    title: Optional[str] = None
    body: str
    severity: Optional[str] = None
    order_idx: int = 0
    # Кто добавил вывод: имя/должность/компания — снимки на момент создания
    created_by_name: Optional[str] = None
    created_by_title: Optional[str] = None
    created_by_org: Optional[str] = None
    created_at: Optional[datetime] = None


class ESGSwotResponse(BaseModel):
    portfolio_strengths: list[ESGSwotItemBrief] = Field(default_factory=list)
    portfolio_weaknesses: list[ESGSwotItemBrief] = Field(default_factory=list)
    company_items: list[ESGSwotItemBrief] = Field(default_factory=list)
    generated_at: datetime


class ESGSwotUpsert(BaseModel):
    id: Optional[UUID] = None
    kind: str = Field(..., pattern="^(strength|weakness)$")
    scope: str = Field("portfolio", pattern="^(portfolio|company)$")
    company_id: Optional[UUID] = None
    title: Optional[str] = Field(None, max_length=255)
    body: str = Field(..., min_length=1)
    severity: Optional[str] = None
    order_idx: int = 0


# =====================================================================
# ESG-отчёты по годам (годовая таблица в профиле зрелости, с 2021)
# =====================================================================

class ESGReportBrief(BaseModel):
    id: Optional[UUID] = None
    company_id: UUID
    year: int
    status: Optional[str] = None        # описание/стандарт отчёта
    report_url: Optional[str] = None
    note: Optional[str] = None
    changed_by_name: Optional[str] = None
    updated_at: Optional[datetime] = None


class ESGReportListResponse(BaseModel):
    company_id: UUID
    company_code: Optional[str] = None
    company_name: Optional[str] = None
    items: list[ESGReportBrief] = Field(default_factory=list)
    last_changed_by_name: Optional[str] = None
    last_changed_at: Optional[datetime] = None
    last_changed_year: Optional[int] = None
    generated_at: datetime


class ESGReportUpsert(BaseModel):
    company_id: UUID
    year: int = Field(..., ge=2000, le=2100)
    status: Optional[str] = Field(None, max_length=255)
    report_url: Optional[str] = Field(None, max_length=2000)
    note: Optional[str] = None


# =====================================================================
# ESG-релевантные KPI по компаниям (подтягиваются из модуля KPI по контексту)
# =====================================================================

class ESGKpiBrief(BaseModel):
    name: str
    unit: Optional[str] = None
    manager: Optional[str] = None
    plan: Optional[float] = None
    fact: Optional[float] = None
    pct: Optional[float] = None       # выполнение, % (с учётом direction)
    direction: str = "up"


class ESGKpiCompany(BaseModel):
    company_id: UUID
    company_code: Optional[str] = None
    kpis: list[ESGKpiBrief] = Field(default_factory=list)


class ESGKpiResponse(BaseModel):
    year: int
    items: list[ESGKpiCompany] = Field(default_factory=list)
    generated_at: datetime


class ESGKpiManagerBrief(BaseModel):
    """Существующая «должность» (менеджер KPI) компании за год — для выбора."""
    id: UUID
    title: str
    short_title: Optional[str] = None


class ESGKpiCreate(BaseModel):
    """Ручное добавление ESG-KPI из дашборда → пишется в модуль KPI (sync с /kpi)."""
    company_id: UUID
    year: int = Field(..., ge=2000, le=2100)
    name: str = Field(..., min_length=1, max_length=500)
    unit: Optional[str] = Field(None, max_length=64)
    direction: str = Field("up", pattern="^(up|down)$")
    plan: Optional[float] = None
    fact: Optional[float] = None
    manager_id: Optional[UUID] = None   # существующая должность; None → общий ESG-менеджер
